Flamethrower rework

Superlagg opened the pull request:
<!-- The text between the arrows are comments - they will not be visible on your PR. -->
<!-- To automatically tag this PR, add the uppercase label(s) surrounded by brackets below, for example: [LABEL] -->

## About the PR <!-- Describe the Pull Request here. What does it change? What other things could this impact? -->

Reworks the flamethrower almost completely.

Changes include:
- The flamethrower is now technically a gun. Doesn't really mean anything to players, but hey.
- Instead of gathering a list of turfs and injecting chems and gas into each of them, it uses a projectile to do this.
  - The projectile is loaded with a set amount of reagents and gas.
  - Every turf it enters, it applies a TOUCH reaction and copies a percentage of its chems into that turf and all of its contents
  - If the flamethrower is LIT, it also sets all the reagents in those atoms to that temperature.
  - This is roughly how the flamethrower used to do it, just that now it does this through a projectile.
- Flamethrowers have three firemodes:
  - A quick, short-ranged 5-spray burst
  - A slow, mid-ranged 3-spritz burst
  - A long-ish ranged semi-auto shot
- Backpack flamethrowers are overall more accurate.
- Spray temperature and spray amount can be configured on the flamethrower.
- Fuel and gas tanks can be swapped out by just clicking it on the flamethrower.
- Adds a variant of the sniper-wallpierce component that only pierces mobs and non-opaque blob tiles.
- Flamethrowers can now be used in gun components.

## Why's this needed? <!-- Describe why you think this should be added to the game. -->

I wanted to see if I could add Tarm's kickin-rad drag-auto component to flamethrowers. Ended up with me reworking them into guns... and not using the component because it did some weird things after I messed with it (It does *not* like being altered or removed after init!).

Also ended up with a really cool-looking flamethrower that actually feels a lot more flamethrower-y. It sprays fire and chems all over the place, burns the heck out of anything that can burn, and is just as dangerous to the user as it is to everything around them. Also roasts the blob real nicely, too.

## Changelog
<!-- If necessary, put your changelog entry below. Otherwise, please delete it.
Use however you want to be credited in the changelog in place of CodeDude.
Use (*) for major changes and (+) for minor changes. For example: -->

```changelog
(u)Superlagg:
(*)Made flamethrowers more configurable and shoot chem-loaded projectiles.
(+)Flamethrowers can now be added to gun components.
```

zjdtmkhzt commented:
Does this mean it will now work in a gun component? :owo:

Superlagg commented:
> Does this mean it will now work in a gun component? :owo:

It does now! :owo:

ZeWaka commented:
A video of basic usage would be great :)

ZeWaka commented on code/WorkInProgress/MechanicMadness.dm:
```diff
@@ -2714,7 +2714,7 @@
 	attackby(obj/item/W as obj, mob/user as mob)
 		if(..(W, user)) return 1
-		else if(istype(W, src.compatible_guns))
+		else if(istype(W, src.compatible_guns) || istype(W, /obj/item/gun/flamethrower))
```
why not add to the list

ZeWaka replied:
oh i see, it's an istype. that's super weird lol

Superlagg replied:
And now its a list!

ZeWaka commented on code/obj/item/flamethrower.dm:
```diff
@@ -2,10 +2,20 @@
 */
-
-/obj/item/flamethrower/
+#define FLAMER_DEFAULT_TEMP 700
+#define FLAMER_MIN_TEMP T0C
+#define FLAMER_MAX_TEMP 1000
```
CELCIUS? KELVIN?

Superlagg replied:
kelvius, of course.

changed it to 1000C

ZeWaka commented on code/obj/item/flamethrower.dm:
```diff
@@ -20,28 +30,133 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
+				return FALSE
+		return TRUE
+
+		// else if(IN_RANGE(get_turf(src), get_turf(src.fueltank), 1))
+		// 	return TRUE
```
?

Superlagg replied:
Must've missed that. And all the others. Removing!

ZeWaka commented on code/obj/item/flamethrower.dm:
```diff
@@ -154,44 +265,76 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
-			return gastank?.reagents
-		else
-			boutput(user, "<span class='alert'>You need to be wearing this flamer's fuelpack to fire it!</span>")
+	// get_reagsource(mob/user)
+	// 	if(gastank in user.get_equipped_items())
+	// 		return gastank?.reagents
+	// 	else
+	// 		boutput(user, "<span class='alert'>You need to be wearing this flamer's fuelpack to fire it!</span>")
```
?

ZeWaka commented on code/obj/item/flamethrower.dm:
```diff
@@ -154,44 +265,76 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
+
+/obj/item/gun/flamethrower/assembled/disposing()
+
+	//src.welder = null
```
?

ZeWaka commented on code/obj/item/flamethrower.dm:
```diff
@@ -515,205 +684,124 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
-	if (src.gastank)
-		dat += text("<br>Air Tank Pressure: [MIXTURE_PRESSURE(src.gastank.air_contents)] (<A HREF='?src=\ref[src];removeair=1'>Remove Air Tank</A>)<BR>")
+	var/dat = "<TT><B>Flamethrower - <A HREF='?src=\ref[src];light=1'>[lit ? "<font color='red'>Lit</font>" : "Unlit"]</a></B><BR>"
+	// var/fueltemp = src.fueltank?.reagents ? (src.fueltank.reagents.total_temperature - T0C) : "NaN"
```
?

Superlagg commented:
https://www.youtube.com/watch?v=ONOEFnOu_ag

Tarmunora reviewed (changes requested):
First batch of review

I can't say I'm a fan on giving the tactical flamer the popup interface - the goal for it was to be as streamlined and, well, idiotproof as possible.

Superlagg commented on code/modules/projectiles/special.dm:
```diff
@@ -963,3 +963,127 @@ ABSTRACT_TYPE(/datum/projectile/special)
 		explosion_new(projectile, get_turf(hit), explosion_power, 1)
 		..()
+
+/datum/projectile/special/shotchem // how do i shot chem
+	name = "chemical bolt"
+	sname = "chembolt"
```
The projectile usually contains chemicals and isn't always on fire. I could have it change its name if its lit, though.

Superlagg commented on code/obj/item/flamethrower.dm:
```diff
@@ -20,28 +30,130 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
+		var/datum/reagents/chems = new(chem_amount)
+		FR.trans_to_direct(chems, chem_amount)
+		SD["crossed_turfs"] = list()
+		SD["FT_reagents"] = chems
```
Its so I can have the reagents stored in the special_data react the chems inside the projectile itself to ensure a fireflash if there's going to be one. Not the best idea, but fireflashes started being more consistent after doing this. I'll find a better solution.

Superlagg replied:
Also turns out storing the reagents in a physical object exposes the reagents to any hotspots on the turf, which is a problem since the reagents stored in the projectile are supposed to be more of a template. Or the for loop is counting the projectile as part of the turf's contents and setting its temperature. Either way, gotta protect the chems. Probably exclude probjectiles from the loop.

Superlagg commented on code/obj/item/gun/gun_parent.dm:
```diff
@@ -49,6 +49,15 @@ var/list/forensic_IDs = new/list() //Global list of all guns, based on bioholder
 
 	var/muzzle_flash = null //set to a different icon state name if you want a different muzzle flash when fired, flash anims located in icons/mob/mob.dmi
 
+	/// Use the fancy shoot manager to handle rapid-fire?
+	var/use_shootloop = 0
+	/// Is the shootloop shootlooping?
+	var/shooting = 0
+	/// Deciseconds the shootloop waits to resume shootlooping after shooting
+	var/refire_delay = 4 DECI SECONDS
+	/// Number of times to have the shootloop loop while shooting
+	var/burst_count = 1
+
```
I never liked how burstfire behavior is defined by the projectile. Very messy if you want a weapon to have multiple different firemodes, and frequently leaves unusable shots in the shot-holder. This ensures that if a shot will be fired if there's enough stuff left in the flamer to fire.

Superlagg commented on code/datums/components/sniper_wallpierce.dm:
```diff
@@ -1,21 +1,27 @@
+	var/only_mobs = 0 // Only pierce mobs
 
-/datum/component/sniper_wallpierce/Initialize(var/num_pierces)
+/datum/component/sniper_wallpierce/Initialize(var/num_pierces, var/_only_mobs)
 	if(!istype(parent, /obj/projectile))
 		return COMPONENT_INCOMPATIBLE
 	if(num_pierces)
 		src.pierces_left=num_pierces
+		src.only_mobs = _only_mobs
 	RegisterSignal(parent, list(COMSIG_PROJ_COLLIDE), .proc/update_pierces)
 
 /datum/component/sniper_wallpierce/proc/update_pierces(var/obj/projectile/P, var/atom/hit)
 	var/turf/T = get_turf(hit)
 	if(isrestrictedz(T.z))
 		return 0
-	if(isrwall(hit) || istype(hit, /obj/machinery/door/poddoor/blast))
-		pierces_left-- //cost an extra pierce for rwalls and blast doors
-	if(pierces_left-- > 0)
-		return PROJ_PASSWALL | PROJ_PASSOBJ
+	if(src.only_mobs)
+		if((ismob(hit) || (istype(hit, /obj/blob) && !hit.opacity)) && pierces_left-- > 0)
+			return PROJ_ATOM_PASSTHROGH
+	else
+		if(isrwall(hit) || istype(hit, /obj/machinery/door/poddoor/blast))
+			pierces_left-- //cost an extra pierce for rwalls and blast doors
+		if(pierces_left-- > 0)
+			return PROJ_PASSWALL | PROJ_PASSOBJ
```
k

Superlagg commented on code/obj/item/flamethrower.dm:
```diff
@@ -20,28 +30,130 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
+		if(!canshoot())
+			return
+
+		var/obj/projectile/SD = P.special_data
```
forgot that part, used to reference the projectile itself before I just made it reference the part I'm interested in.

Superlagg commented on code/obj/item/flamethrower.dm:
```diff
@@ -20,28 +30,130 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
+		var/turf/T = get_turf(src)
+		var/datum/gas_mixture/airgas = unpool(/datum/gas_mixture)
+		airgas.volume = 1
+		airgas.temperature = SD["burn_temp"]
+		airgas.merge(GTAC.remove_ratio(rem_ratio * 0.5))
```
It shouldnt be setting the temp at all if the projectile isnt lit. Fixed that.

Superlagg commented on code/obj/item/flamethrower.dm:
```diff
@@ -390,106 +489,155 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
+			qdel(S)
+			boutput(user, "<span class='notice'>The igniter is now unsecured!</span>")
+
+	else if (istype(src, /obj/item/gun/flamethrower/backtank))
```
Turns out that backtank stuff wasn't needed, it was being handled in the UI thing.

Superlagg commented on code/modules/projectiles/special.dm:
```diff
@@ -963,3 +963,127 @@ ABSTRACT_TYPE(/datum/projectile/special)
+
+
+	/// Releases some of the projectile's gas into the turf
+	proc/splurt_gas(turf/T, all_of_it = 0)
```
schplordge

Superlagg commented:
> I can't say I'm a fan on giving the tactical flamer the popup interface - the goal for it was to be as streamlined and, well, idiotproof as possible.

While I get that it's meant to be simple to use for Nukies, and that most of the settings don't matter for them either... yeah maybe its better to just remove the menu and lock the firemode to something between wide and narrow spray.

Superlagg commented on code/obj/item/flamethrower.dm:
```diff
@@ -20,28 +30,130 @@ GETLINEEEEEEEEEEEEEEEEEEEEE
+		else if(src.fueltank?.reagents.total_volume <= 0)
+			boutput(user, "<span class='alert'>This [src]'s fuel source is empty!</span>")
+			return FALSE
+		else if(T_env && src.gastank?.air_contents)
```
Now it does!

Carbadox commented:
looks fucking sick, hell yeah

Tarmunora commented on code/datums/components/sniper_wallpierce.dm:
```diff
@@ -2,27 +2,43 @@
+
+/datum/component/pierce_mobs/UnregisterFromParent()
+	UnregisterSignal(parent, COMSIG_PROJ_COLLIDE)
+	. = ..()
```
Move to new file

Superlagg replied:
Done

Tarmunora commented on code/obj/item/flamethrower.dm:
```diff
@@ -2,10 +2,26 @@
 */
-
-/obj/item/flamethrower/
+#define FLAMER_DEFAULT_TEMP 700 KELVIN
+#define FLAMER_BACKTANK_TEMP 1000 KELVIN + T0C
+#define FLAMER_MIN_TEMP T0C
+#define FLAMER_MAX_TEMP 1000 KELVIN + T0C
+#define FLAMER_DEFAULT_CHEM_AMT 10
+#define FLAMER_BACKTANK_CHEM_AMT 25
+#define FLAMER_MIN_CHEM_AMT 5
+#define FLAMER_MAX_CHEM_AMT 20
+#define FLAMER_MODE_LOOSE 1
+#define FLAMER_MODE_TIGHT 2
+#define FLAMER_MODE_SINGLE 3
+#define FLAMER_MODE_BACKTANK 4
```
if these are not used elsewhere, undef them at the end of the file

Superlagg replied:
Done!

Tarmunora commented on code/datums/components/sniper_wallpierce.dm:
```diff
@@ -1,5 +1,6 @@
 /datum/component/sniper_wallpierce
 	var/pierces_left = 1 //default to 1 wall
+	var/only_mobs = 0 // Only pierce mobs
```
```suggestion
```

Superlagg replied:
darn, forgot about that

Adharainspace commented on _std/defines/projectiles.dm:
```diff
@@ -4,7 +4,7 @@
 
 //pass flags on hit thing take priotity if they exist
-#define PROJ_ATOM_PASSTHROGH	(1<<4)
+#define PROJ_ATOM_PASSTHROUGH	(1<<4)
```
thanks for fixing this typo. my standard library review blessing is now upon you

Superlagg replied:
np =3

Superlagg commented:
Aight, made some changes:
- The first two firemodes set the flamer to use fullauto, and the last one to not fullauto.
- Added sound effects to setting changes, and made removing the tanks not close the setting window.
- Removed my shootmanager thing. That'll be a later PR.
- Changed the max amount of reagents the flamer can load to 100, and the default to 20.
- Reduced the amount of gas loaded per shot. Was making semiauto run out of gas long before emptying the reagents.
- Changed the sound to a shorter, slightly less shrill version of its old sound.
- Gun and projectiles no longer spam the chat with useless messages.

Tarmunora commented on code/modules/projectiles/special.dm:
```diff
@@ -963,3 +963,118 @@ ABSTRACT_TYPE(/datum/projectile/special)
+			if(istype(A, /obj/overlay) || istype(A, /obj/projectile))
+				continue
+			copied.reaction(A, TOUCH, 0, 0)
+			if(!A.reagents)
+				A.create_reagents(100)
+			copied.copy_to(A.reagents, 1, copy_temperature = 1)
+			if(O.special_data["IS_LIT"]) // Heat if needed
+				A.reagents.set_reagent_temp(O.special_data["burn_temp"], TRUE)
```
I feel like a sweeping rework like this is perhaps a good time to get rid of flamers coping into mobs/objects, and only apply the touch reactions (including possible skin penetration) to those

Superlagg replied:
I dunno about that. Flamers have always done that, and in a way became one of its main features. Though, yeah, it could be used to dupe chems, many chemnerds have come to expect it to do... whatever it is chemnerds do with it.

ZeWaka replied:
fuk da nerds

Tarmunora replied:
indeed

Superlagg commented on code/modules/projectiles/projectile_parent.dm:
```diff
@@ -517,6 +517,9 @@
 		return
 
+	temperature_expose(datum/gas_mixture/air, exposed_temperature, exposed_volume)
+		return
```
I assume this would prevent the flamer projectile from heating up its contents when setting everything in its turf on fire, yeah?

Tarmunora replied:
that's the hope
